uievents/mouse/cancel-mousedown-in-subframe.html WPT is failing

aprotyas opened the pull request:
#### e837c341a756a244e94b2b2cdd166aa02b92e217
<pre>
uievents/mouse/cancel-mousedown-in-subframe.html WPT is failing
<a href="https://bugs.webkit.org/show_bug.cgi?id=262691">https://bugs.webkit.org/show_bug.cgi?id=262691</a>
rdar://116516596

Reviewed by Wenson Hsieh.

This patch changes how we handle mouse events following preventDefault
on mousedown events. Following this patch, if we initiate a drag from a
cancelled mousedown event, we keep sending events to the originating
frame till the terminating mouseup corresponding to said drag gesture is
dispatched.

We do so by slightly adjusting our logic around the
`EventHandler::capturesDragging` state, which represents whether or not
we can capture the element under a mouse event during a drag.
Previously, for a mousedown event that was preventDefault-ed,
`capturesDragging` would be unconditionally set to false for the event
handler associated with said frame, which meant that we would not be
capturing the element under the mouse event in said frame, which
resulted in the bug (and the WPT failure).

Instead, this commit changes `capturesDragging` from a boolean flag to a
structure with a little more information. Namely, we keep track of the
reason why drag capture is not possible. This allows us to say that
`capturesDragging` is not possible _because_ the mousePress in question is
cancelled (`MousePressIsCancelled`). The main frame event handler can now
capture the right element when successfully passing a mouse press event to
a subframe. This is represented either as `localSubframe-&gt;ownerElement()`
or as the new boolean state `m_isCapturingRootElementForMouseEvent`;
since we now capture an element to keep sending events to when either
`capturesDragging` is true or when the reason drag capture is not possible is
`MousePressIsCancelled`.

This patch aligns our behavior with the Gecko engine, and gets us to pass
the `uievents/mouse/cancel-mousedown-in-subframe.html` web platform test.

* LayoutTests/TestExpectations:

Adjust test expectations now that we&apos;re passing the WPT.

* LayoutTests/fast/events/cancel-mousedown-and-drag-from-frame-to-other-frame-expected.txt: Added.
* LayoutTests/fast/events/cancel-mousedown-and-drag-from-frame-to-other-frame.html: Renamed from LayoutTests/fast/events/mouse-drag-from-frame-to-other-frame.html.
* LayoutTests/fast/events/cancel-mousedown-and-drag-to-frame-expected.txt: Added.
* LayoutTests/fast/events/cancel-mousedown-and-drag-to-frame.html: Renamed from LayoutTests/fast/events/mouse-drag-from-frame.html.
* LayoutTests/fast/events/mouse-drag-from-frame-expected.txt: Removed.
* LayoutTests/fast/events/mouse-drag-from-frame-to-other-frame-expected.txt: Removed.
* LayoutTests/fast/events/resources/mouse-drag-from-frame-subframe.html:
* LayoutTests/fast/events/resources/mouse-drag-from-frame-target-subframe.html:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/cancel-mousedown-in-subframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/cancel-mousedown-in-subframe.html:

Uncomment two assertion statements. This was probably an import mistake.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:

Update two layout tests fast/events/mouse-drag-from-frame.html and
fast/events/mouse-drag-from-frame-to-other-frame.html to reflect the new
event retention behavior change introduced by this PR.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchMouseEvent):
* Source/WebCore/dom/Element.h:

Introduce an overload for `dispatchMouseEvent` that accepts an outparam
`isEventCancelled` that a caller can use to learn if the provided mouse
event is preventDefault-ed.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::clear):
(WebCore::EventHandler::handleMousePressEvent):
(WebCore::EventHandler::handleMouseDoubleClickEvent):
(WebCore::EventHandler::handleMouseMoveEvent):
(WebCore::EventHandler::handleMouseReleaseEvent):
(WebCore::EventHandler::clearDragState):
(WebCore::EventHandler::setCapturingMouseEventsElement):
(WebCore::EventHandler::dispatchMouseEvent):

Event retention behavior adjustment as described in the commit message
above.

* Source/WebCore/page/EventHandler.h:
(WebCore::EventHandler::CapturesDragging::CapturesDragging):
(WebCore::EventHandler::CapturesDragging::operator=):
(WebCore::EventHandler::CapturesDragging::operator bool const):
(WebCore::EventHandler::CapturesDragging::inabilityReason const):
(WebCore::EventHandler::capturesDragging const):
(WebCore::EventHandler::isCapturingMouseEventsElement const):
(WebCore::EventHandler::resetCapturingMouseEventsElement):

Update the capturesDragging variable to act as a boolean flag but also
contain a reason if capturesDragging::operator bool() returns false.
Moreover, introduce an additional boolean state that indicates when
we&apos;re capturing the root HTML element, and not some subframe, for this
event retention behavior. This is necessary so we can make a distinction
between a null element and the root element.

Canonical link: <a href="https://commits.webkit.org/269246@main">https://commits.webkit.org/269246@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ccd37fb8cfc14a40408d9a96d9bd3e1e46a021ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22017 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22229 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23086 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23901 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20382 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22263 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26477 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22546 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21448 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22248 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21899 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24753 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19010 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19954 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26211 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20032 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20179 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24078 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20609 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17547 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19763 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5246 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24168 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20559 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
From the commit message

> that accepts an outparam `isEventCancelled` that a callee can use to learn

this should read _caller_ instead, will fix.

whsieh commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -530,7 +530,43 @@ class EventHandler {
+        }
+
+        CapturesDragging(DragCaptureInabilityReason inabilityReason)
+        : m_state { unexpect, inabilityReason }
```
Nit - indentation
```suggestion
            : m_state { unexpect, inabilityReason }
```

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/pull/18685/commits/40172157018f802cb724d0a46a30b22367de5f55

whsieh commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -467,6 +467,12 @@ static ShouldIgnoreMouseEvent dispatchPointerEventIfNeeded(Element& element, con
+    return dispatchMouseEvent(platformEvent, eventType, isEventDefaultPrevented, detail, relatedTarget, isSyntheticClick);
+}
+
+bool Element::dispatchMouseEvent(const PlatformMouseEvent& platformEvent, const AtomString& eventType,  bool& isEventDefaultPrevented, int detail, Element* relatedTarget, IsSyntheticClick isSyntheticClick)
```
Out-params usually go at the end of a function.

I think you could also consider returning a `std::pair` (with an enum class indicating whether or not the page prevented default). This probably requires mechanically refactoring the many call sites of `dispatchMouseEvent` though, so we can do this in a separate change.

aprotyas replied:
> I think you could also consider returning a std::pair (with an enum class indicating whether or not the page prevented default). This probably requires mechanically refactoring the many call sites of dispatchMouseEvent though, so we can do this in a separate change.

Yeah I think that would be more ergonomic. I added a FIXME for that: https://bugs.webkit.org/show_bug.cgi?id=263034

aprotyas replied:
> Out-params usually go at the end of a function.

Fixed in https://github.com/WebKit/WebKit/commit/40172157018f802cb724d0a46a30b22367de5f55

whsieh reviewed (approved):
Is it worth rebaselining/adjusting the existing layout tests instead of just removing them?

It seems we might be losing a bit of coverage w.r.t. where `mousemove` events are dispatched.

aprotyas committed:
uievents/mouse/cancel-mousedown-in-subframe.html WPT is failing
https://bugs.webkit.org/show_bug.cgi?id=262691
rdar://116516596

Reviewed by Wenson Hsieh.

This patch changes how we handle mouse events following preventDefault
on mousedown events. Following this patch, if we initiate a drag from a
cancelled mousedown event, we keep sending events to the originating
frame till the terminating mouseup corresponding to said drag gesture is
dispatched.

We do so by slightly adjusting our logic around the
`EventHandler::capturesDragging` state, which represents whether or not
we can capture the element under a mouse event during a drag.
Previously, for a mousedown event that was preventDefault-ed,
`capturesDragging` would be unconditionally set to false for the event
handler associated with said frame, which meant that we would not be
capturing the element under the mouse event in said frame, which
resulted in the bug (and the WPT failure).

Instead, this commit changes `capturesDragging` from a boolean flag to a
structure with a little more information. Namely, we keep track of the
reason why drag capture is not possible. This allows us to say that
`capturesDragging` is not possible _because_ the mousePress in question is
cancelled (`MousePressIsCancelled`). The main frame event handler can now
capture the right element when successfully passing a mouse press event to
a subframe. This is represented either as `localSubframe->ownerElement()`
or as the new boolean state `m_isCapturingRootElementForMouseEvent`;
since we now capture an element to keep sending events to when either
`capturesDragging` is true or when the reason drag capture is not possible is
`MousePressIsCancelled`.

This patch aligns our behavior with the Gecko engine, and gets us to pass
the `uievents/mouse/cancel-mousedown-in-subframe.html` web platform test.

* LayoutTests/TestExpectations:

Adjust test expectations now that we're passing the WPT.

* LayoutTests/fast/events/cancel-mousedown-and-drag-from-frame-to-other-frame-expected.txt: Added.
* LayoutTests/fast/events/cancel-mousedown-and-drag-from-frame-to-other-frame.html: Renamed from LayoutTests/fast/events/mouse-drag-from-frame-to-other-frame.html.
* LayoutTests/fast/events/cancel-mousedown-and-drag-to-frame-expected.txt: Added.
* LayoutTests/fast/events/cancel-mousedown-and-drag-to-frame.html: Renamed from LayoutTests/fast/events/mouse-drag-from-frame.html.
* LayoutTests/fast/events/mouse-drag-from-frame-expected.txt: Removed.
* LayoutTests/fast/events/mouse-drag-from-frame-to-other-frame-expected.txt: Removed.
* LayoutTests/fast/events/resources/mouse-drag-from-frame-subframe.html:
* LayoutTests/fast/events/resources/mouse-drag-from-frame-target-subframe.html:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/cancel-mousedown-in-subframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/cancel-mousedown-in-subframe.html:

Uncomment two assertion statements. This was probably an import mistake.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:

Update two layout tests fast/events/mouse-drag-from-frame.html and
fast/events/mouse-drag-from-frame-to-other-frame.html to reflect the new
event retention behavior change introduced by this PR.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchMouseEvent):
* Source/WebCore/dom/Element.h:

Introduce an overload for `dispatchMouseEvent` that accepts an outparam
`isEventCancelled` that a caller can use to learn if the provided mouse
event is preventDefault-ed.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::clear):
(WebCore::EventHandler::handleMousePressEvent):
(WebCore::EventHandler::handleMouseDoubleClickEvent):
(WebCore::EventHandler::handleMouseMoveEvent):
(WebCore::EventHandler::handleMouseReleaseEvent):
(WebCore::EventHandler::clearDragState):
(WebCore::EventHandler::setCapturingMouseEventsElement):
(WebCore::EventHandler::dispatchMouseEvent):

Event retention behavior adjustment as described in the commit message
above.

* Source/WebCore/page/EventHandler.h:
(WebCore::EventHandler::CapturesDragging::CapturesDragging):
(WebCore::EventHandler::CapturesDragging::operator=):
(WebCore::EventHandler::CapturesDragging::operator bool const):
(WebCore::EventHandler::CapturesDragging::inabilityReason const):
(WebCore::EventHandler::capturesDragging const):
(WebCore::EventHandler::isCapturingMouseEventsElement const):
(WebCore::EventHandler::resetCapturingMouseEventsElement):

Update the capturesDragging variable to act as a boolean flag but also
contain a reason if capturesDragging::operator bool() returns false.
Moreover, introduce an additional boolean state that indicates when
we're capturing the root HTML element, and not some subframe, for this
event retention behavior. This is necessary so we can make a distinction
between a null element and the root element.

Canonical link: https://commits.webkit.org/269246@main
